nemo/collections/tts/models/tacotron2.py

```diff
@@ -0,0 +1,282 @@
+        )
+
+        loss = self.loss(
+            mel_out=spec_dec,
+            mel_out_postnet=spec_postnet,
+            gate_out=gate,
+            mel_target=spec,
+            gate_target=gate_padded,
+            target_len=spec_len,
+            pad_value=self.pad_value,
+        )
+        return {
+            "loss": loss,
+            "mel_target": spec,
+            "mel_postnet": spec_postnet,
+            "gate": gate,
+            "gate_target": gate_padded,
+            "alignments": alignments,
+        }
+
+    def validation_epoch_end(self, outputs):
+        tacotron2_log_to_tb_func(
+            self.logger.experiment, outputs[0].values(), self.global_step, tag="eval", log_images=True, add_audio=False
+        )
+        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
+        tensorboard_logs = {'val_loss': avg_loss}
+        return {'val_loss': avg_loss, 'log': tensorboard_logs}
+
+    def __setup_dataloader_from_config(self, cfg, shuffle_should_be: bool = True, name: str = "train"):
+        if "dataset" not in cfg or not isinstance(cfg.dataset, DictConfig):
+            raise ValueError(f"No dataset for {name}")  # TODO
+        if "dataloader_params" not in cfg or not isinstance(cfg.dataloader_params, DictConfig):
+            raise ValueError(f"No dataloder_params for {name}")  # TODO
+        if shuffle_should_be:
+            if 'shuffle' not in cfg.dataloader_params:
+                logging.warning(
+                    f"Shuffle should be set to True for {self}'s {name} dataloader but was not found in its "
+                    "config. Manually setting to True"
+                )
+                with open_dict(cfg.dataloader_params):
+                    cfg.dataloader_params.shuffle = True
+            elif not cfg.dataloader_params.shuffle:
+                logging.error(f"The {name} dataloader for {self} has shuffle set to False!!!")
+        elif not shuffle_should_be and cfg.dataloader_params.shuffle:
+            logging.error(f"The {name} dataloader for {self} has shuffle set to True!!!")
+
+        labels = cfg.dataset.params.labels
+        dataset = Tacotron2Model.from_config_dict(
+            cfg.dataset, bos_id=len(labels), eos_id=len(labels) + 1, pad_id=len(labels) + 2,
+        )
+        return torch.utils.data.DataLoader(dataset, collate_fn=dataset.collate_fn, **cfg.dataloader_params)
+
+    def setup_training_data(self, cfg):
+        self._train_dl = self.__setup_dataloader_from_config(cfg)
+
+    def setup_validation_data(self, cfg):
+        self._validation_dl = self.__setup_dataloader_from_config(cfg, shuffle_should_be=False, name="validation")
+
+    @classmethod
+    def list_available_models(cls) -> 'Optional[Dict[str, str]]':
+        """TODO: Implement me!"""
+        pass
+
+    @classmethod
+    def from_pretrained(cls, name: str):
+        """TODO: Implement me!"""
+        pass
+
+    def export(self, **kwargs):
```

Comment:
Please remove this method (you can do so if you rebase/merge latest candidate into your branch)

Reply:
Done in next commit.

Reply:
Undone